pkg/service/restore/service.go

```diff
@@ -0,0 +1,280 @@
+		"task_id", taskID,
+		"run_id", runID,
+	)
+
+	w, err := s.newWorker(ctx, clusterID)
+	if err != nil {
+		return errors.Wrap(err, "create worker")
+	}
+	w.setRunInfo(taskID, runID)
+
+	if err := w.initTarget(ctx, properties); err != nil {
+		return errors.Wrap(err, "init target")
+	}
+
+	if w.target.Continue {
+		if err := w.decorateWithPrevRun(ctx); err != nil {
+			return err
+		}
+	}
+
+	if w.run.PrevID == uuid.Nil {
+		// Reset metrics on fresh start
+		w.metrics.ResetClusterMetrics(w.run.ClusterID)
+		// Cache must be initialised only once, as it contains information already lost
+		// in the cluster (e.g. tombstone_gc mode, views definition, etc).
+		if err := w.initUnits(ctx); err != nil {
+			return errors.Wrap(err, "initialize units")
+		}
+		if err := w.initViews(ctx); err != nil {
+			return errors.Wrap(err, "initialize views")
+		}
+	} else {
+		w.clonePrevProgress(ctx)
+		// Check that all units are still present after resume
+		for _, u := range w.run.Units {
+			for _, t := range u.Tables {
+				ok, err := w.client.TableExists(ctx, "", u.Keyspace, t.Table)
+				if err != nil {
+					return errors.Wrapf(err, "query table %s.%s existence", u.Keyspace, t.Table)
+				}
+				if !ok {
+					return errors.Wrapf(err, "table %s.%s doesn't exist", u.Keyspace, t.Table)
+				}
+			}
+		}
+	}
+	w.insertRun(ctx)
+
+	if w.target.RestoreTables {
+		var totalBytesToRestore int64
+		for _, unit := range w.run.Units {
+			totalBytesToRestore += unit.Size
+		}
+		tw := newTablesWorker(w, s.repairSvc, totalBytesToRestore)
+		err = tw.restore(ctx)
+	} else {
+		sw := &schemaWorker{worker: w}
+		err = sw.restore(ctx)
+	}
+
+	if err == nil {
+		w.run.Stage = StageDone
+		w.insertRun(ctx)
+	}
+	return err
+}
+
+// GetTarget returns validated target from properties.
+func (s *Service) GetTarget(ctx context.Context, clusterID uuid.UUID, properties json.RawMessage) (Target, error) {
+	w, err := s.newWorker(ctx, clusterID)
+	if err != nil {
+		return Target{}, errors.Wrap(err, "create worker")
+	}
+	defer w.clusterSession.Close()
+
+	if err := w.initTarget(ctx, properties); err != nil {
+		return Target{}, errors.Wrap(err, "initialize target")
+	}
+	return w.target, nil
+}
+
+// GetUnits returns validated units from properties.
+func (s *Service) GetUnits(ctx context.Context, clusterID uuid.UUID, properties json.RawMessage) ([]Unit, error) {
+	w, err := s.newWorker(ctx, clusterID)
+	if err != nil {
+		return nil, errors.Wrap(err, "create worker")
+	}
+	defer w.clusterSession.Close()
+
+	if err := w.initTarget(ctx, properties); err != nil {
+		return nil, errors.Wrap(err, "initialize target")
+	}
+	if err := w.initUnits(ctx); err != nil {
+		return nil, errors.Wrap(err, "initialize units")
+	}
+	return w.run.Units, nil
+}
+
+// GetTargetUnitsViews returns all information necessary for task validation and --dry-run.
+func (s *Service) GetTargetUnitsViews(ctx context.Context, clusterID uuid.UUID, properties json.RawMessage) (Target, []Unit, []View, error) {
```

Comment:
This public method (part of the interface) is almost not tested.
Please test it (migrate the tests listed in comments under GetTarget and GetUnits)